File: core/src/tables/cblc.rs
use std::ops::Range;

use crate::parser::*;
use crate::{Error, Result};

#[derive(Clone, Copy, Debug)]
struct BitmapFlags(u8);

impl FromData for BitmapFlags {
    const NAME: &'static str = "BitFlags";

    #[inline]
    fn parse(data: &[u8]) -> Result<Self> {
        Ok(BitmapFlags(data[0]))
    }
}

impl std::fmt::Display for BitmapFlags {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use std::fmt::Write;
        let mut s = String::new();

        let bits = U8Bits(self.0);
        writeln!(&mut s, "{}", bits)?;
        if bits[0] { s.push_str("Bit 0: Horizontal\n"); }
        if bits[1] { s.push_str("Bit 1: Vertical\n"); }

        s.pop(); // pop last new line

        f.write_str(&s)
    }
}

struct SubtableArray {
    offset: usize,
    num_of_subtables: u32,
}

struct SubtableInfo {
    first_glyph: GlyphId,
    last_glyph: GlyphId,
    offset: usize,
}

pub fn parse(parser: &mut Parser) -> Result<()> {
    let start = parser.offset();

    let major_version = parser.read::<u16>("Major version")?;
    let minor_version = parser.read::<u16>("Minor version")?;
    // Some old Noto Emoji fonts have a 2.0 version.
    if !((major_version == 2 || major_version == 3) && minor_version == 0) {
        return Err(Error::InvalidTableVersion);
    }

    let num_sizes = parser.read::<u32>("Number of tables")?;

    let mut subtable_arrays = Vec::new();
    for _ in 0..num_sizes {
        parser.begin_group("Table");

        let offset = parser.read::<Offset32>("Offset to index subtable")?.to_usize();
        parser.read::<u32>("Index tables size")?;
        let num_of_subtables = parser.read::<u32>("Number of index subtables")?;
        parser.read::<u32>("Reserved")?;

        parser.begin_group("Line metrics for horizontal text");
        parse_line_metrics(parser)?;
        parser.end_group();

        parser.begin_group("Line metrics for vertical text");
        parse_line_metrics(parser)?;
        parser.end_group();

        parser.read::<GlyphId>("Lowest glyph index")?;
        parser.read::<GlyphId>("Highest glyph index")?;
        parser.read::<u8>("Horizontal pixels per em")?;
        parser.read::<u8>("Vertical pixels per em")?;
        parser.read::<u8>("Bit depth")?;
        parser.read::<BitmapFlags>("Flags")?;

        parser.end_group();

        subtable_arrays.push(SubtableArray { offset, num_of_subtables });
    }

    subtable_arrays.sort_by_key(|v| v.offset);
    subtable_arrays.dedup_by_key(|v| v.offset);

    let mut subtables = Vec::new();
    for array in subtable_arrays {
        parser.jump_to(start + array.offset)?;

        for _ in 0..array.num_of_subtables {
            parser.begin_group("Index subtable array");
            let first_glyph = parser.read::<GlyphId>("First glyph ID")?;
            let last_glyph = parser.read::<GlyphId>("Last glyph ID")?;
            let offset2 = parser.read::<Offset32>("Additional offset to index subtable")?.to_usize();
            parser.end_group();

            subtables.push(SubtableInfo {
                first_glyph,
                last_glyph,
                offset: start + array.offset + offset2,
            });
        }
    }

    subtables.sort_by_key(|v| v.offset);
    subtables.dedup_by_key(|v| v.offset);

    for info in subtables {
        parser.jump_to(info.offset)?;
        parser.begin_group("Index subtable");
        let index_format = parser.read::<u16>("Index format")?;
        parser.read::<u16>("Image format")?;
        parser.read::<Offset32>("Offset to image data")?;

        match index_format {
            1 => {
                // TODO: check
                let count = info.last_glyph.0 - info.first_glyph.0 + 2;
                parser.read_array::<Offset32>("Offsets", "Offset", count as usize)?;
            }
            2 => {
                parser.read::<u32>("Image size")?;
                parse_big_glyph_metrics(parser)?;
            }
            3 => {
                // TODO: check
                let count = info.last_glyph.0 - info.first_glyph.0 + 2;
                parser.read_array::<Offset16>("Offsets", "Offset", count as usize)?;
            }
            4 => {
                let num_glyphs = parser.read::<u32>("Number of glyphs")?;
                for _ in 0..=num_glyphs {
                    parser.read::<GlyphId>("Glyph ID")?;
                    parser.read::<Offset16>("Offset")?;
                }
            }
            5 => {
                parser.read::<u32>("Image size")?;
                parse_big_glyph_metrics(parser)?;
                let num_glyphs = parser.read::<u32>("Number of glyphs")?;
                parser.read_array::<GlyphId>("Glyphs", "Glyph ID", num_glyphs as usize)?;
            }
            _ => return Err(Error::InvalidValue),
        }

        parser.end_group();
    }

    Ok(())
}

fn parse_line_metrics(parser: &mut Parser) -> Result<()> {
    parser.read::<i8>("Ascender")?;
    parser.read::<i8>("Descender")?;
    parser.read::<u8>("Max width")?;
    parser.read::<i8>("Caret slope numerator")?;
    parser.read::<i8>("Caret slope denominator")?;
    parser.read::<i8>("Caret offset")?;
    parser.read::<i8>("Min origin SB")?;
    parser.read::<i8>("Min advance SB")?;
    parser.read::<i8>("Max before BL")?;
    parser.read::<i8>("Min after BL")?;
    parser.read_bytes(2, "Padding")?;
    Ok(())
}

pub fn parse_small_glyph_metrics(parser: &mut Parser) -> Result<()> {
    parser.read::<u8>("Height")?;
    parser.read::<u8>("Width")?;
    parser.read::<i8>("X-axis bearing")?;
    parser.read::<i8>("Y-axis bearing")?;
    parser.read::<u8>("Advance")?;
    Ok(())
}

pub fn parse_big_glyph_metrics(parser: &mut Parser) -> Result<()> {
    parser.read::<u8>("Height")?;
    parser.read::<u8>("Width")?;
    parser.read::<i8>("Horizontal X-axis bearing")?;
    parser.read::<i8>("Horizontal Y-axis bearing")?;
    parser.read::<u8>("Horizontal advance")?;
    parser.read::<i8>("Vertical X-axis bearing")?;
    parser.read::<i8>("Vertical Y-axis bearing")?;
    parser.read::<u8>("Vertical advance")?;
    Ok(())
}


pub struct CblcIndex {
    pub image_format: u16,
    pub range: Range<usize>,
}

pub fn collect_indices(data: &[u8]) -> Result<Vec<CblcIndex>> {
    let mut parser = SimpleParser::new(data);
    let mut locations = Vec::new();

    let start = parser.offset();

    parser.skip::<u16>(); // Major version
    parser.skip::<u16>(); // Minor version

    let num_sizes = parser.read::<u32>()?;

    let mut subtable_arrays = Vec::new();
    for _ in 0..num_sizes {
        let offset = parser.read::<Offset32>()?.to_usize();
        parser.skip::<u32>(); // Index tables size
        let num_of_subtables = parser.read::<u32>()?;
        parser.advance(36);

        subtable_arrays.push(SubtableArray { offset, num_of_subtables });
    }

    subtable_arrays.sort_by_key(|v| v.offset);
    subtable_arrays.dedup_by_key(|v| v.offset);

    let mut subtables = Vec::new();
    for array in subtable_arrays {
        parser.jump_to(start + array.offset)?;

        for _ in 0..array.num_of_subtables {
            let first_glyph = parser.read::<GlyphId>()?;
            let last_glyph = parser.read::<GlyphId>()?;
            let offset2 = parser.read::<Offset32>()?.to_usize();

            subtables.push(SubtableInfo {
                first_glyph,
                last_glyph,
                offset: start + array.offset + offset2
            });
        }
    }

    subtables.sort_by_key(|v| v.offset);
    subtables.dedup_by_key(|v| v.offset);

    for info in subtables {
        parser.jump_to(info.offset)?;
        let index_format = parser.read::<u16>()?;
        let image_format = parser.read::<u16>()?;
        let image_data_offset = parser.read::<Offset32>()?.to_usize();

        match index_format {
            1 => {
                let count = info.last_glyph.0 - info.first_glyph.0 + 2;
                let mut offsets = Vec::new();
                for _ in 0..count {
                    let offset = parser.read::<Offset32>()?.to_usize();
                    offsets.push(image_data_offset + offset);
                }

                offsets.sort();
                offsets.dedup();

                for i in 0..offsets.len()-1 {
                    let start = offsets[i];
                    let end = offsets[i + 1];
                    locations.push(CblcIndex { image_format, range: start..end });
                }
            }
            2 => {
                let image_size = parser.read::<u32>()? as usize;

                let count = info.last_glyph.0 - info.first_glyph.0 + 1;
                let mut offset = image_data_offset;
                for _ in 0..count {
                    let start = offset;
                    offset += image_size;
                    let end = offset;
                    locations.push(CblcIndex { image_format, range: start..end });
                }
            }
            3 => {
                let count = info.last_glyph.0 - info.first_glyph.0 + 2;
                let mut offsets = Vec::new();
                for _ in 0..count {
                    let offset = parser.read::<Offset16>()?.to_usize();
                    offsets.push(image_data_offset + offset);
                }

                offsets.sort();
                offsets.dedup();

                for i in 0..offsets.len() - 1 {
                    let start = offsets[i];
                    let end = offsets[i + 1];
                    locations.push(CblcIndex { image_format, range: start..end });
                }
            }
            4 => {
                let num_glyphs = parser.read::<u32>()?;
                let mut offsets = Vec::new();
                for _ in 0..=num_glyphs {
                    parser.skip::<GlyphId>();
                    let offset = parser.read::<Offset16>()?.to_usize();
                    offsets.push(image_data_offset + offset);
                }

                offsets.sort();
                offsets.dedup();

                for i in 0..offsets.len() - 1 {
                    let start = offsets[i];
                    let end = offsets[i + 1];
                    locations.push(CblcIndex { image_format, range: start..end });
                }
            }
            5 => {
                let image_size = parser.read::<u32>()? as usize;
                parser.advance(8); // big metrics
                let num_glyphs = parser.read::<u32>()?;

                let mut offset = image_data_offset;
                let mut offsets = Vec::new();
                for _ in 0..=num_glyphs {
                    offsets.push(offset);
                    offset += image_size;
                }

                offsets.sort();
                offsets.dedup();

                for i in 0..offsets.len() - 1 {
                    let start = offsets[i];
                    let end = offsets[i + 1];
                    locations.push(CblcIndex { image_format, range: start..end });
                }
            }
            _ => return Err(Error::InvalidValue),
        }
    }

    locations.sort_by_key(|v| v.range.start);

    Ok(locations)
}

File: src/tables/cblc.cpp
#include <bitset>

#include "src/algo.h"
#include "src/parser.h"
#include "tables.h"

struct EblcBitmapFlags
{
    static const int Size = 1;
    static const QString Type;

    static EblcBitmapFlags parse(const quint8 *data)
    {
        return { data[0] };
    }

    static QString toString(const EblcBitmapFlags &value)
    {
        std::bitset<8> bits(value.d);
        auto flagsStr = QString::fromUtf8(bits.to_string().c_str()) + '\n';

        if (bits[0]) flagsStr += "Bit 0: Horizontal\n";
        if (bits[1]) flagsStr += "Bit 1: Vertical\n";

        flagsStr.chop(1); // trim trailing newline
        return flagsStr;
    }

    quint8 d;
};

const QString EblcBitmapFlags::Type = "BitFlags";

static void parseSbitLineMetrics(Parser &parser)
{
    parser.read<Int8>("Ascender");
    parser.read<Int8>("Descender");
    parser.read<UInt8>("Max width");
    parser.read<Int8>("Caret slope numerator");
    parser.read<Int8>("Caret slope denominator");
    parser.read<Int8>("Caret offset");
    parser.read<Int8>("Min origin SB");
    parser.read<Int8>("Min advance SB");
    parser.read<Int8>("Max before BL");
    parser.read<Int8>("Min after BL");
    parser.readBytes(2, "Padding");
}

void parseSbitSmallGlyphMetrics(Parser &parser)
{
    parser.read<UInt8>("Height");
    parser.read<UInt8>("Width");
    parser.read<Int8>("X-axis bearing");
    parser.read<Int8>("Y-axis bearing");
    parser.read<UInt8>("Advance");
}

void parseSbitBigGlyphMetrics(Parser &parser)
{
    parser.read<UInt8>("Height");
    parser.read<UInt8>("Width");
    parser.read<Int8>("Horizontal X-axis bearing");
    parser.read<Int8>("Horizontal Y-axis bearing");
    parser.read<UInt8>("Horizontal advance");
    parser.read<Int8>("Vertical X-axis bearing");
    parser.read<Int8>("Vertical Y-axis bearing");
    parser.read<UInt8>("Vertical advance");
}

void parseCblc(Parser &parser)
{
    const auto start = parser.offset();

    const auto majorVersion = parser.read<UInt16>("Major version");
    const auto minorVersion = parser.read<UInt16>("Minor version");
    // Some old Noto Emoji fonts have a 2.0 version.
    if (!((majorVersion == 2 || majorVersion == 3) && minorVersion == 0)) {
        throw "invalid table version";
    }

    const auto numSizes = parser.read<UInt32>("Number of tables");

    struct SubtableArray
    {
        quint32 offset;
        quint32 numOfSubtables;
    };

    QVector<SubtableArray> subtableArrays;
    for (uint i = 0; i < numSizes; ++i) {
        parser.beginGroup("Table");

        const auto offset = parser.read<Offset32>("Offset to index subtable");
        parser.read<UInt32>("Index tables size");
        const auto numOfSubtables = parser.read<UInt32>("Number of index subtables");
        parser.read<UInt32>("Reserved");

        parser.beginGroup("Line metrics for horizontal text");
        parseSbitLineMetrics(parser);
        parser.endGroup();

        parser.beginGroup("Line metrics for vertical text");
        parseSbitLineMetrics(parser);
        parser.endGroup();

        parser.read<GlyphId>("Lowest glyph index");
        parser.read<GlyphId>("Highest glyph index");
        parser.read<UInt8>("Horizontal pixels per em");
        parser.read<UInt8>("Vertical pixels per em");
        parser.read<UInt8>("Bit depth");
        parser.read<EblcBitmapFlags>("Flags");

        parser.endGroup();

        subtableArrays.append(SubtableArray { offset, numOfSubtables });
    }

    algo::sort_all_by_key(subtableArrays, &SubtableArray::offset);
    algo::dedup_vector_by_key(subtableArrays, &SubtableArray::offset);

    struct SubtableInfo
    {
        quint16 firstGlyph;
        quint16 lastGlyph;
        quint32 offset;
    };

    QVector<SubtableInfo> subtables;
    for (const auto array : subtableArrays) {
        parser.jumpTo(start + array.offset);

        for (uint i = 0; i < array.numOfSubtables; ++i) {
            parser.beginGroup("Index subtable array");
            const auto firstGlyph = parser.read<GlyphId>("First glyph ID");
            const auto lastGlyph = parser.read<GlyphId>("Last glyph ID");
            const auto offset2 = parser.read<Offset32>("Additional offset to index subtable");
            parser.endGroup();

            subtables.append(SubtableInfo { firstGlyph, lastGlyph, start + array.offset + offset2 });
        }
    }

    algo::sort_all_by_key(subtables, &SubtableInfo::offset);
    algo::dedup_vector_by_key(subtables, &SubtableInfo::offset);

    for (const auto info : subtables) {
        parser.jumpTo(info.offset);
        parser.beginGroup("Index subtable");
        const auto indexFormat = parser.read<UInt16>("Index format");
        parser.read<UInt16>("Image format");
        parser.read<Offset32>("Offset to image data");

        if (indexFormat == 1) {
            // TODO: check
            const auto count = info.lastGlyph - info.firstGlyph + 2;
            parser.readArray<Offset32>("Offsets", "Offset", quint32(count));
        } else if (indexFormat == 2) {
            parser.read<UInt32>("Image size");
            parseSbitBigGlyphMetrics(parser);
        } else if (indexFormat == 3) {
            // TODO: check
            const auto count = info.lastGlyph - info.firstGlyph + 2;
            parser.readArray<Offset16>("Offsets", "Offset", quint32(count));
        } else if (indexFormat == 4) {
            const auto numGlyphs = parser.read<UInt32>("Number of glyphs");
            for (uint i = 0; i <= numGlyphs; ++i) {
                parser.read<GlyphId>("Glyph ID");
                parser.read<Offset16>("Offset");
            }
        } else if (indexFormat == 5) {
            parser.read<UInt32>("Image size");
            parseSbitBigGlyphMetrics(parser);
            const auto numGlyphs = parser.read<UInt32>("Number of glyphs");
            parser.readArray<GlyphId>("Glyphs", "Glyph ID", numGlyphs);
        } else {
            throw "unsupported index format";
        }

        parser.endGroup();
    }
}

QVector<CblcIndex> parseCblcLocations(ShadowParser parser)
{
    QVector<CblcIndex> locations;

    const auto start = parser.offset();

    parser.skip<UInt16>(); // Major version
    parser.skip<UInt16>(); // Minor version

    const auto numSizes = parser.read<UInt32>();

    struct SubtableArray
    {
        quint32 offset;
        quint32 numOfSubtables;
    };

    QVector<SubtableArray> subtableArrays;
    for (uint i = 0; i < numSizes; ++i) {
        const auto offset = parser.read<Offset32>();
        parser.skip<UInt32>(); // Index tables size
        const auto numOfSubtables = parser.read<UInt32>();
        parser.advance(36);

        subtableArrays.append(SubtableArray { offset, numOfSubtables });
    }

    algo::sort_all_by_key(subtableArrays, &SubtableArray::offset);
    algo::dedup_vector_by_key(subtableArrays, &SubtableArray::offset);

    struct SubtableInfo
    {
        quint16 firstGlyph;
        quint16 lastGlyph;
        quint32 offset;
    };

    QVector<SubtableInfo> subtables;
    for (const auto array : subtableArrays) {
        parser.jumpTo(start + array.offset);

        for (uint i = 0; i < array.numOfSubtables; ++i) {
            const auto firstGlyph = parser.read<GlyphId>();
            const auto lastGlyph = parser.read<GlyphId>();
            const auto offset2 = parser.read<Offset32>();

            subtables.append(SubtableInfo { firstGlyph, lastGlyph, start + array.offset + offset2 });
        }
    }

    algo::sort_all_by_key(subtables, &SubtableInfo::offset);
    algo::dedup_vector_by_key(subtables, &SubtableInfo::offset);

    for (const auto info : subtables) {
        parser.jumpTo(info.offset);
        const auto indexFormat = parser.read<UInt16>();
        const auto imageFormat = parser.read<UInt16>();
        const auto imageDataOffset = parser.read<Offset32>();

        if (indexFormat == 1) {
            const auto count = quint32(info.lastGlyph - info.firstGlyph + 2);
            QVector<quint32> offsets;
            for (quint32 i = 0; i < count; ++i) {
                offsets.append(imageDataOffset + parser.read<Offset32>());
            }

            algo::sort_all(offsets);
            algo::dedup_vector(offsets);

            for (int i = 0; i < offsets.size() - 1; ++i) {
                const auto start = offsets.at(i);
                const auto end = offsets.at(i + 1);
                locations.append(CblcIndex { imageFormat, Range { start, end } });
            }
        } else if (indexFormat == 2) {
            const auto imageSize = parser.read<UInt32>();

            const auto count = quint32(info.lastGlyph - info.firstGlyph + 1);
            quint32 offset = imageDataOffset;
            for (quint32 i = 0; i < count; ++i) {
                const auto start = offset;
                offset += imageSize;
                const auto end = offset;
                locations.append(CblcIndex { imageFormat, Range { start, end } });
            }
        } else if (indexFormat == 3) {
            const auto count = quint32(info.lastGlyph - info.firstGlyph + 2);
            QVector<quint32> offsets;
            for (quint32 i = 0; i < count; ++i) {
                offsets.append(imageDataOffset + parser.read<Offset16>());
            }

            algo::sort_all(offsets);
            algo::dedup_vector(offsets);

            for (int i = 0; i < offsets.size() - 1; ++i) {
                const auto start = offsets.at(i);
                const auto end = offsets.at(i + 1);
                locations.append(CblcIndex { imageFormat, Range { start, end } });
            }
        } else if (indexFormat == 4) {
            const auto numGlyphs = parser.read<UInt32>();
            QVector<quint32> offsets;
            for (quint32 i = 0; i <= numGlyphs; ++i) {
                parser.skip<GlyphId>();
                offsets.append(imageDataOffset + parser.read<Offset16>());
            }

            algo::sort_all(offsets);
            algo::dedup_vector(offsets);

            for (int i = 0; i < offsets.size() - 1; ++i) {
                const auto start = offsets.at(i);
                const auto end = offsets.at(i + 1);
                locations.append(CblcIndex { imageFormat, Range { start, end } });
            }
        } else if (indexFormat == 5) {
            const auto imageSize = parser.read<UInt32>();
            parser.advance(8); // big metrics
            const auto numGlyphs = parser.read<UInt32>();

            quint32 offset = imageDataOffset;
            QVector<quint32> offsets;
            for (quint32 i = 0; i <= numGlyphs; ++i) {
                offsets.append(offset);
                offset += imageSize;
            }

            algo::sort_all(offsets);
            algo::dedup_vector(offsets);

            for (int i = 0; i < offsets.size() - 1; ++i) {
                const auto start = offsets.at(i);
                const auto end = offsets.at(i + 1);
                locations.append(CblcIndex { imageFormat, Range { start, end } });
            }
        }
    }

    algo::sort_all(locations, [](const auto &a, const auto &b) {
        return a.range.start < b.range.start;
    });

    return locations;
}
